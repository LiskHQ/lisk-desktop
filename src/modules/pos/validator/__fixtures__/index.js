export * from './mockValidators';
export * from './mockSentVotes';
export * from './mockReceivedVotes';
export * from './mockUnlocks';
export * from './mockForgers';
export * from './mockGenerator';
