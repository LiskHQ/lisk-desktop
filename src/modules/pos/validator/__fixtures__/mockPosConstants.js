export const mockPosConstants = {
  data: {
    factorSelfStakes: 10,
    maxLengthName: 20,
    maxNumberSentStakes: 10,
    maxNumberPendingUnlocks: 20,
    failSafeMissedBlocks: 50,
    failSafeInactiveWindow: 260000,
    punishmentWindow: 780000,
    roundLength: 103,
    minWeightStandby: '100000000000',
    numberActiveValidators: 101,
    numberStandbyValidators: 2,
    posTokenID: '0000000000000000',
    maxBFTWeightCap: 500,
    commissionIncreasePeriod: 260000,
    maxCommissionIncreaseRate: 500,
    extraCommandFees: {
      validatorRegistrationFee: '1000000000',
    },
  },
  meta: {},
};
