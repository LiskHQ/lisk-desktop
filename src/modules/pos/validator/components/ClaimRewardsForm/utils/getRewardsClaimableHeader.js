import grid from 'flexboxgrid/dist/flexboxgrid.css';

const getRewardsClaimableHeader = (t) => [
  {
    title: t && t('Name'),
    classList: `${grid['col-xs-4']}`,
  },
  {
    title: t && t('Token'),
    classList: `${grid['col-xs-2']}`,
  },
  {
    title: t && t('Reward amount'),
    classList: `${grid['col-xs-4']}`,
  },
  {
    title: t && t('Fiat'),
    classList: `${grid['col-xs-2']}`,
  },
];

export default getRewardsClaimableHeader;
