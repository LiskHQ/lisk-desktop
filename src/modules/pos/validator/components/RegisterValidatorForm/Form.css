@import '../../../../../../setup/react/app/mixins.css';

.wrapper {
  box-sizing: border-box;
  width: 100%;
  padding: var(--horizontal-modal-padding-l);
  padding-bottom: 0;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;

  & .header {
    & h2 {
      color: var(--color-black);
      font-size: var(--font-size-h3);
      font-weight: var(--font-weight-bold);
      margin: 0;
    }
  }

  & .container {
    padding: 24px 0;

    & > .description {
      @mixin contentLargest;

      color: var(--color-slate-gray);
      margin: 0;
      padding: 0;
    }

    & > .description:nth-of-type(1) {
      margin-top: 2px;
    }

    & > .description:nth-of-type(2) {
      margin-top: 25px;
    }

    & > .link {
      @mixin contentLargest bold;

      color: var(--color-ultramarine-blue);
      margin-top: 18px;
      text-decoration: none;
    }

    & > .label {
      @mixin contentNormal bold;

      color: var(--color-maastricht-blue);
      margin: 16px 0 8px;

      &:first-child {
        margin-top: 0;
      }
    }

    & .input {
      padding: 8px 23px 8px 14px;
    }

    & > footer {
      margin-top: 50px;
    }
  }
}
