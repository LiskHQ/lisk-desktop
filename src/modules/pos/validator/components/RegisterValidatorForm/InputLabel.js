import React from 'react';
import Tooltip from 'src/theme/Tooltip';
import styles from './Form.css';

const InputLabel = ({ title, tooltip }) => (
  <label className={styles.label}>
    {title}
    <Tooltip position="right">
      <p>
        {tooltip}
      </p>
    </Tooltip>
  </label>
);

export default InputLabel;
