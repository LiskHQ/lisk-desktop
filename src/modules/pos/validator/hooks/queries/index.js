export * from './useValidators';
export * from './useReceivedVotes';
export * from './useSentVotes';
export * from './useUnlocks';
export * from './useForgersValidator';
export * from './usePosConstants';
