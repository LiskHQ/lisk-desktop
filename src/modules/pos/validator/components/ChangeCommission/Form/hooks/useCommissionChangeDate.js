import { usePosConstants } from '@pos/validator/hooks/queries';
import { useCurrentAccount } from '@account/hooks';
import { useTransactions } from '@transaction/hooks/queries';
import { MODULE_COMMANDS_NAME_MAP } from '@transaction/configuration/moduleCommand';
import { useNetworkStatus } from '@network/hooks/queries';

const getDateWhenCommissionCanBeIncreased = (
  timestampCommissionChange,
  commissionIncreasePeriod,
  blockTime
) => {
  const timestampWhenCommissionCanBeIncreased =
    (timestampCommissionChange + commissionIncreasePeriod * blockTime) * 1000;

  return new Date(timestampWhenCommissionCanBeIncreased);
};

export const useCommissionChangeDate = () => {
  const { data: posConstants } = usePosConstants();
  const [
    {
      metadata: { address },
    },
  ] = useCurrentAccount();
  const { data: transactions, isLoading } = useTransactions({
    config: {
      params: {
        address,
        moduleCommand: MODULE_COMMANDS_NAME_MAP.changeCommission,
        limit: 1,
        sort: 'timestamp:desc',
      },
    },
  });

  const lastChangeCommissionTimestamp = transactions?.data[0]?.block?.timestamp;
  const { data: networkStatus } = useNetworkStatus();

  return {
    date:
      lastChangeCommissionTimestamp &&
      getDateWhenCommissionCanBeIncreased(
        lastChangeCommissionTimestamp,
        posConstants.data?.commissionIncreasePeriod,
        networkStatus?.data?.genesis?.blockTime
      ),
    isLoading,
  };
};
