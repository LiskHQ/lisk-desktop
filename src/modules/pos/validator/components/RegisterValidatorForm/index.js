import React from 'react';
import { useTranslation } from 'react-i18next';
import { MODULE_COMMANDS_NAME_MAP } from 'src/modules/transaction/configuration/moduleCommand';
import BoxHeader from 'src/theme/box/header';
import BoxContent from 'src/theme/box/content';
import { Input } from 'src/theme';
import TxComposer from '@transaction/components/TxComposer';
import useValidatorName from '../../hooks/useValidatorName';
import useValidatorKey from '../../hooks/useValidatorKey';
import InputLabel from './InputLabel';
import styles from './Form.css';
import { usePosConstants } from '../../hooks/queries';
import usePosToken from '../../hooks/usePosToken';

const isFormValid = (name, generatorKey, blsKey, proofOfPossession) =>
  !name.error && !name.loading && !generatorKey.error && !blsKey.error && !proofOfPossession.error;

const getTooltips = (field, t) => {
  if (field === 'name') {
    return t('Max. 20 characters, a-z, 0-1, no special characters except !@$_.');
  }
  return t('Run lisk keys:generate and copy the value of {{field}}', { field });
};

// eslint-disable-next-line max-statements
const RegisterValidatorForm = ({ nextStep, prevState }) => {
  const { t } = useTranslation();
  const { params } = prevState?.transactionJSON || {};
  const [name, setName] = useValidatorName(params?.name);
  const [generatorKey, setGenKey] = useValidatorKey(
    'generatorKey',
    t('Please enter a valid generator key value'),
    params?.generatorKey
  );
  const [blsKey, setBlsKey] = useValidatorKey(
    'blsKey',
    t('Please enter a valid bls key value'),
    params?.blsKey
  );
  const [proofOfPossession, setPop] = useValidatorKey(
    'proofOfPossession',
    t('Please enter a valid proof of possession value'),
    params?.proofOfPossession
  );

  const { data: posConstants } = usePosConstants();
  const { token } = usePosToken();

  const onConfirm = (formProps, transactionJSON, selectedPriority, fees) => {
    nextStep({
      selectedPriority,
      formProps,
      transactionJSON,
      fees,
    });
  };

  const onChangeUsername = ({ target: { value } }) => {
    setName(value);
  };

  const registerValidatorFormProps = {
    isFormValid: isFormValid(name, generatorKey, blsKey, proofOfPossession),
    moduleCommand: MODULE_COMMANDS_NAME_MAP.registerValidator,
    fields: { token },
    extraCommandFee: posConstants?.data?.extraCommandFees?.validatorRegistrationFee || 0,
  };
  const commandParams = {
    name: name.value,
    blsKey: blsKey.value,
    generatorKey: generatorKey.value,
    proofOfPossession: proofOfPossession.value,
  };

  return (
    <section className={styles.wrapper}>
      <TxComposer
        onConfirm={onConfirm}
        formProps={registerValidatorFormProps}
        commandParams={commandParams}
        buttonTitle="Go to confirmation"
      >
        <>
          <BoxHeader className={styles.header}>
            <h2>{t('Register validator')}</h2>
          </BoxHeader>
          <BoxContent className={`${styles.container} select-name-container`}>
            <InputLabel title={t('Your name')} tooltip={getTooltips('name', t)} />
            <div className={styles.inputContainer}>
              <Input
                data-name="validator-username"
                autoComplete="off"
                onChange={onChangeUsername}
                name="validator-username"
                value={name.value}
                placeholder={t('e.g. peter_pan')}
                className={`${styles.input} select-name-input`}
                error={name.error && name.value.length}
                isLoading={name.loading}
                status={name.error && name.value.length ? 'error' : 'ok'}
                feedback={name.message}
              />
            </div>

            <InputLabel title={t('Generator key')} tooltip={getTooltips('generatorKey', t)} />
            <div className={styles.inputContainer}>
              <Input
                data-name="generator-publicKey"
                autoComplete="off"
                onChange={(e) => setGenKey(e.target.value)}
                name="generatorKey"
                value={generatorKey.value}
                placeholder={t('A string value')}
                className={`${styles.input} generator-publicKey-input`}
                error={generatorKey.error}
                status={generatorKey.error ? 'error' : 'ok'}
                feedback={generatorKey.message}
              />
            </div>

            <InputLabel title={t('BLS Key')} tooltip={getTooltips('blsKey', t)} />
            <div className={styles.inputContainer}>
              <Input
                data-name="bls-key"
                autoComplete="off"
                onChange={(e) => setBlsKey(e.target.value)}
                name="blsKey"
                value={blsKey.value}
                placeholder={t('A string value')}
                className={`${styles.input} bls-key-input`}
                error={blsKey.error}
                status={blsKey.error ? 'error' : 'ok'}
                feedback={blsKey.message}
              />
            </div>

            <InputLabel
              title={t('BLS Proof Of Possession')}
              tooltip={getTooltips('proofOfPossession', t)}
            />
            <div className={styles.inputContainer}>
              <Input
                data-name="pop"
                autoComplete="off"
                onChange={(e) => setPop(e.target.value)}
                name="proofOfPossession"
                value={proofOfPossession.value}
                placeholder={t('A string value')}
                className={`${styles.input} pop-input`}
                error={proofOfPossession.error}
                status={proofOfPossession.error ? 'error' : 'ok'}
                feedback={proofOfPossession.message}
              />
            </div>
          </BoxContent>
        </>
      </TxComposer>
    </section>
  );
};

export default RegisterValidatorForm;
