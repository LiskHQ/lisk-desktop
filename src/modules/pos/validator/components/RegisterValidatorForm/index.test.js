import { mountWithQueryClient } from 'src/utils/testHelpers';
import * as keys from '@tests/constants/keys';
import { useTokensBalance } from '@token/fungible/hooks/queries';
import mockSavedAccounts from '@tests/fixtures/accounts';
import { mockAuth } from 'src/modules/auth/__fixtures__';

import { mockTokensBalance } from 'src/modules/token/fungible/__fixtures__';
import { usePosConstants } from '../../hooks/queries';
import { mockPosConstants } from '../../__fixtures__/mockPosConstants';
import useValidatorName from '../../hooks/useValidatorName';
import useValidatorKey from '../../hooks/useValidatorKey';
import RegisterValidatorForm from '.';

const mockCurrentAccount = mockSavedAccounts[0];

jest.mock('../../hooks/queries');
jest.mock('@token/fungible/hooks/queries');
jest.mock('@account/hooks/useCurrentAccount', () => ({
  useCurrentAccount: jest.fn(() => [mockCurrentAccount]),
}));
jest.mock('../../hooks/useValidatorName', () => jest.fn());
jest.mock('../../hooks/useValidatorKey', () => jest.fn());
jest.mock('@account/hooks/useDeprecatedAccount', () => ({
  useDeprecatedAccount: jest.fn().mockReturnValue({
    isSuccess: true,
    isLoading: false,
  }),
}));
jest.mock('@auth/hooks/queries', () => ({
  ...jest.requireActual('@auth/hooks/queries'),
  useAuth: jest.fn().mockReturnValue({ data: mockAuth }),
}));

const genKey = {
  value: keys.genKey,
  error: false,
  message: '',
};
const blsKey = {
  value: keys.blsKey,
  error: false,
  message: '',
};
const pop = {
  value: keys.pop,
  error: false,
  message: '',
};
const empty = {
  value: '',
  error: true,
  message: 'Can not be empty',
};

jest.useFakeTimers();

describe('RegisterValidatorForm', () => {
  const props = {
    prevState: {},
    nextStep: jest.fn(),
  };

  const setName = jest.fn();
  const setKey = jest.fn();

  usePosConstants.mockReturnValue({ data: mockPosConstants });
  useTokensBalance.mockReturnValue({ data: mockTokensBalance, isLoading: false });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Renders correctly', () => {
    it('renders properly RegisterValidatorForm component', () => {
      useValidatorName.mockReturnValue([empty, setName]);
      useValidatorKey.mockReturnValue([empty, setKey]);
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);
      expect(wrapper).toContainMatchingElement('.select-name-container');
      expect(wrapper).toContainMatchingElement('.select-name-input');
      expect(wrapper).toContainMatchingElement('.feedback');
      expect(wrapper).toContainMatchingElement('.confirm-btn');
      expect(wrapper.find('button.confirm-btn')).toBeDisabled();
    });

    it('type a valid and unused username', async () => {
      useValidatorName.mockReturnValue([empty, setName]);
      useValidatorKey.mockReturnValue([empty, setKey]);
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);

      wrapper
        .find('input.select-name-input')
        .simulate('change', { target: { value: 'myvalidator' } });
      wrapper
        .find('input.generator-publicKey-input')
        .simulate('change', { target: { value: genKey.value, name: 'generatorKey' } });
      wrapper
        .find('input.bls-key-input')
        .simulate('change', { target: { value: blsKey.value, name: 'blsKey' } });
      wrapper
        .find('input.pop-input')
        .simulate('change', { target: { value: pop.value, name: 'proofOfPossession' } });
      expect(setName).toHaveBeenCalledTimes(1);
      expect(setKey).toHaveBeenCalledWith(genKey.value);
      expect(setKey).toHaveBeenCalledWith(blsKey.value);
      expect(setKey).toHaveBeenCalledWith(pop.value);
    });
  });

  describe('Handles errors', () => {
    const wrongName = {
      value: 's',
      error: true,
      message: 'Too short',
    };
    const validName = {
      value: 'some_name',
      error: false,
      message: '',
    };
    const wrongKey = {
      value: 'wrong_value',
      error: true,
      message: 'Invalid hex value',
    };

    it('Display validator name input errors', () => {
      useValidatorKey.mockReturnValueOnce([genKey, setKey]);
      useValidatorKey.mockReturnValueOnce([blsKey, setKey]);
      useValidatorKey.mockReturnValueOnce([pop, setKey]);
      useValidatorName.mockReturnValue([wrongName, setName]); // invalid
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);
      expect(wrapper.find('button.confirm-btn')).toBeDisabled();
      expect(wrapper.find('.feedback').at(0)).toHaveText('Too short');
    });

    it('Display generator key input error', () => {
      useValidatorKey.mockReturnValueOnce([wrongKey, setKey]);
      useValidatorKey.mockReturnValueOnce([blsKey, setKey]);
      useValidatorKey.mockReturnValueOnce([pop, setKey]);
      useValidatorName.mockReturnValue([validName, setName]); // valid
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);
      expect(wrapper.find('button.confirm-btn')).toBeDisabled();
      expect(wrapper.find('.feedback').at(1)).toHaveText('Invalid hex value');
    });

    it('Display bls key input error', () => {
      useValidatorKey.mockReturnValueOnce([genKey, setKey]);
      useValidatorKey.mockReturnValueOnce([wrongKey, setKey]);
      useValidatorKey.mockReturnValueOnce([pop, setKey]);
      useValidatorName.mockReturnValue([validName, setName]); // valid
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);
      expect(wrapper.find('button.confirm-btn')).toBeDisabled();
      expect(wrapper.find('.feedback').at(2)).toHaveText('Invalid hex value');
    });

    it('Display pop key input error', () => {
      useValidatorKey.mockReturnValueOnce([genKey, setKey]);
      useValidatorKey.mockReturnValueOnce([blsKey, setKey]);
      useValidatorKey.mockReturnValueOnce([wrongKey, setKey]);
      useValidatorName.mockReturnValue([validName, setName]); // valid
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);
      expect(wrapper.find('button.confirm-btn')).toBeDisabled();
      expect(wrapper.find('.feedback').at(3)).toHaveText('Invalid hex value');
    });
  });

  describe('Pass valid tx', () => {
    const validName = {
      value: 'some_name',
      error: false,
      message: '',
    };
    const rawTx = {
      transactionJSON: {
        fee: '1000000000',
        module: 'pos',
        command: 'registerValidator',
        nonce: '0',
        params: {
          blsKey: blsKey.value,
          generatorKey: genKey.value,
          proofOfPossession: pop.value,
          name: validName.value,
        },
        signatures: [],
        senderPublicKey: mockCurrentAccount.metadata.pubkey,
      },
      formProps: {
        composedFees: [
          {
            title: 'Transaction',
            value: '0 LSK',
            components: [],
          },
          {
            title: 'Message',
            value: '0 LSK',
            isHidden: true,
            components: [],
          },
        ],
        extraCommandFee: '1000000000',
        fields: {
          token: mockTokensBalance.data[0],
        },
        isFormValid: true,
        moduleCommand: 'pos:registerValidator',
      },
      fees: [
        {
          title: 'Transaction',
          value: '0 LSK',
          components: [],
        },
        {
          title: 'Message',
          value: '0 LSK',
          isHidden: true,
          components: [],
        },
      ],
      selectedPriority: { title: 'Normal', selectedIndex: 0, value: 0 },
    };

    it('accept a valid form', () => {
      useValidatorKey.mockReturnValueOnce([genKey, setKey]);
      useValidatorKey.mockReturnValueOnce([blsKey, setKey]);
      useValidatorKey.mockReturnValueOnce([pop, setKey]);
      useValidatorName.mockReturnValue([validName, setName]); // valid
      const wrapper = mountWithQueryClient(RegisterValidatorForm, props);
      wrapper.find('button.confirm-btn').simulate('click');
      expect(props.nextStep).toHaveBeenCalledWith(rawTx);
    });
  });
});
