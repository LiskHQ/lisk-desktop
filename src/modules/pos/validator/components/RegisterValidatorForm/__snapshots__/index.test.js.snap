// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterValidatorForm Pass valid tx accept a valid form 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "fees": [
          {
            "components": [
              {
                "type": "bytesFee",
                "value": 330000n,
              },
              {
                "type": "Registration",
                "value": 1000000000n,
              },
            ],
            "title": "Transaction",
            "value": "10.0033 LSK",
          },
          {
            "components": [],
            "isHidden": true,
            "title": "Message",
            "value": "0 LSK",
          },
        ],
        "formProps": {
          "composedFees": [
            {
              "components": [
                {
                  "type": "bytesFee",
                  "value": 330000n,
                },
                {
                  "type": "Registration",
                  "value": 1000000000n,
                },
              ],
              "title": "Transaction",
              "value": "10.0033 LSK",
            },
            {
              "components": [],
              "isHidden": true,
              "title": "Message",
              "value": "0 LSK",
            },
          ],
          "extraCommandFee": "1000000000",
          "fields": {
            "token": {
              "availableBalance": 10000000000,
              "baseDenom": "beddows",
              "chainID": "00000001",
              "chainName": "Lisk",
              "denomUnits": [
                {
                  "decimals": 0,
                  "denom": "beddows",
                },
                {
                  "aliases": [
                    "Lisk",
                  ],
                  "decimals": 8,
                  "denom": "lisk",
                },
              ],
              "description": "Base token for the Lisk ecosystem",
              "displayDenom": "lisk",
              "logo": {
                "png": "https://lisk-qa.ams3.digitaloceanspaces.com/Artboard%201%20copy%2019.png",
                "svg": "https://downloads.lisk.com/lisk/images/tokens/lsk.svg",
              },
              "networkType": "mainnet",
              "symbol": "LSK",
              "tokenID": "0000000100000000",
              "tokenName": "Lisk",
            },
          },
          "isFormValid": true,
          "moduleCommand": "pos:registerValidator",
        },
        "selectedPriority": {
          "selectedIndex": 0,
          "title": "Normal",
          "value": 0,
        },
        "transactionJSON": {
          "command": "registerValidator",
          "fee": "1000330000",
          "module": "pos",
          "nonce": "0",
          "params": {
            "blsKey": "830ce8c4a0b4f40b9b2bd2f16e835676b003ae28ec367432af9bfaa4d5201051786643620eff288077c1e7a8415c0285",
            "generatorKey": "dcad7c69505d549803fb6a755e81cdcb0a33ea95b6476e2585149f8a42c9c882",
            "name": "some_name",
            "proofOfPossession": "722b19e4b302e3e13ef097b417b651feadc8e28754530119911561c27b9478cdcd6b7ada331037bbda778b0b325aab5a79f34b31ea780acd01bf67d38268c43ea0ea75a5e757a76165253e1e20680c4cfd884ed63f5663c7b940e67162d5f715",
          },
          "senderPublicKey": "cf434a889d6c7a064e8de61bb01759a76f585e5ff45a78ba8126ca332601f535",
          "signatures": [],
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
